// mendixmodelsdk/src/gen/changedatacapture.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.changedatacapture = exports.StructureVersionInfo = void 0;
const internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
const projects_1 = require("./projects");
/**
 * @ignore
 */
var changedatacapture;
(function (changedatacapture) {
    /**
     * Interfaces and instance classes for types from the Mendix sub meta model `ChangeDataCapture`.
     */
    /**
     * NOTE: This class is experimental and is subject to change in newer Model SDK versions.
     *
     * @ignore
     *
     * In version 11.9.0: introduced
     */
    class EntityChangeProperty extends internal.Element {
        constructor(model, structureTypeName, id, isPartial, unit, container) {
            super(model, structureTypeName, id, isPartial, unit, container);
            /** @internal */
            this.__exposedName = new internal.PrimitiveProperty(EntityChangeProperty, this, "exposedName", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new EntityChangeProperty() cannot be invoked directly, please use 'model.changedatacapture.createEntityChangeProperty()'");
            }
        }
        get containerAsEntityChangeSource() {
            return super.getContainerAs(EntityChangeSource);
        }
        get exposedName() {
            return this.__exposedName.get();
        }
        set exposedName(newValue) {
            this.__exposedName.set(newValue);
        }
        /** @internal */
        _initializeDefaultProperties() {
            super._initializeDefaultProperties();
        }
    }
    EntityChangeProperty.structureTypeName = "ChangeDataCapture$EntityChangeProperty";
    EntityChangeProperty.versionInfo = new exports.StructureVersionInfo({
        introduced: "11.9.0",
        experimental: {
            currentValue: true
        }
    }, internal.StructureType.Element);
    changedatacapture.EntityChangeProperty = EntityChangeProperty;
    /**
     * NOTE: This class is experimental and is subject to change in newer Model SDK versions.
     *
     * @ignore
     *
     * In version 11.9.0: introduced
     */
    class AssociationProperty extends EntityChangeProperty {
        constructor(model, structureTypeName, id, isPartial, unit, container) {
            super(model, structureTypeName, id, isPartial, unit, container);
            /** @internal */
            this.__association = new internal.ByNameReferenceProperty(AssociationProperty, this, "association", null, "DomainModels$AssociationBase");
            if (arguments.length < 4) {
                throw new Error("new AssociationProperty() cannot be invoked directly, please use 'model.changedatacapture.createAssociationProperty()'");
            }
        }
        get containerAsEntityChangeSource() {
            return super.getContainerAs(EntityChangeSource);
        }
        get association() {
            return this.__association.get();
        }
        set association(newValue) {
            this.__association.set(newValue);
        }
        get associationQualifiedName() {
            return this.__association.qualifiedName();
        }
        /**
         * Creates and returns a new AssociationProperty instance in the SDK and on the server.
         * The new AssociationProperty will be automatically stored in the 'properties' property
         * of the parent EntityChangeSource element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  11.9.0 and higher
         */
        static createIn(container) {
            internal.createInVersionCheck(container.model, AssociationProperty.structureTypeName, { start: "11.9.0" });
            return internal.instancehelpers.createElement(container, AssociationProperty, "properties", true);
        }
        /**
         * Creates and returns a new AssociationProperty instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        static create(model) {
            return internal.instancehelpers.createElement(model, AssociationProperty);
        }
        /** @internal */
        _initializeDefaultProperties() {
            super._initializeDefaultProperties();
        }
    }
    AssociationProperty.structureTypeName = "ChangeDataCapture$AssociationProperty";
    AssociationProperty.versionInfo = new exports.StructureVersionInfo({
        introduced: "11.9.0",
        properties: {
            association: {
                required: {
                    currentValue: true
                }
            }
        },
        experimental: {
            currentValue: true
        }
    }, internal.StructureType.Element);
    changedatacapture.AssociationProperty = AssociationProperty;
    /**
     * NOTE: This class is experimental and is subject to change in newer Model SDK versions.
     *
     * @ignore
     *
     * In version 11.9.0: introduced
     */
    class AttributeProperty extends EntityChangeProperty {
        constructor(model, structureTypeName, id, isPartial, unit, container) {
            super(model, structureTypeName, id, isPartial, unit, container);
            /** @internal */
            this.__attribute = new internal.ByNameReferenceProperty(AttributeProperty, this, "attribute", null, "DomainModels$Attribute");
            if (arguments.length < 4) {
                throw new Error("new AttributeProperty() cannot be invoked directly, please use 'model.changedatacapture.createAttributeProperty()'");
            }
        }
        get containerAsEntityChangeSource() {
            return super.getContainerAs(EntityChangeSource);
        }
        get attribute() {
            return this.__attribute.get();
        }
        set attribute(newValue) {
            this.__attribute.set(newValue);
        }
        get attributeQualifiedName() {
            return this.__attribute.qualifiedName();
        }
        /**
         * Creates and returns a new AttributeProperty instance in the SDK and on the server.
         * The new AttributeProperty will be automatically stored in the 'properties' property
         * of the parent EntityChangeSource element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  11.9.0 and higher
         */
        static createIn(container) {
            internal.createInVersionCheck(container.model, AttributeProperty.structureTypeName, { start: "11.9.0" });
            return internal.instancehelpers.createElement(container, AttributeProperty, "properties", true);
        }
        /**
         * Creates and returns a new AttributeProperty instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        static create(model) {
            return internal.instancehelpers.createElement(model, AttributeProperty);
        }
        /** @internal */
        _initializeDefaultProperties() {
            super._initializeDefaultProperties();
        }
    }
    AttributeProperty.structureTypeName = "ChangeDataCapture$AttributeProperty";
    AttributeProperty.versionInfo = new exports.StructureVersionInfo({
        introduced: "11.9.0",
        properties: {
            attribute: {
                required: {
                    currentValue: true
                }
            }
        },
        experimental: {
            currentValue: true
        }
    }, internal.StructureType.Element);
    changedatacapture.AttributeProperty = AttributeProperty;
    /**
     * See: {@link https://docs.mendix.com/refguide/event-streams/#change-data-capture relevant section in reference guide}
     *
     * NOTE: This class is experimental and is subject to change in newer Model SDK versions.
     *
     * @ignore
     *
     * In version 11.5.0: introduced
     */
    class EntityChangeDataCaptureService extends projects_1.projects.Document {
        constructor(model, structureTypeName, id, isPartial, container) {
            super(model, structureTypeName, id, isPartial, container);
            /** @internal */
            this.__sources = new internal.PartListProperty(EntityChangeDataCaptureService, this, "sources", []);
            /** @internal */
            this.__serviceName = new internal.PrimitiveProperty(EntityChangeDataCaptureService, this, "serviceName", "", internal.PrimitiveTypeEnum.String);
            this._containmentName = "documents";
        }
        get containerAsFolderBase() {
            return super.getContainerAs(projects_1.projects.FolderBase);
        }
        /**
         * NOTE: This property is experimental and is subject to change in newer Model SDK versions.
         *
         * @ignore
         */
        get sources() {
            return this.__sources.get();
        }
        /**
         * In version 11.8.0: introduced
         */
        get serviceName() {
            return this.__serviceName.get();
        }
        set serviceName(newValue) {
            this.__serviceName.set(newValue);
        }
        /**
         * Creates a new EntityChangeDataCaptureService unit in the SDK and on the server.
         * Expects one argument, the projects.IFolderBase in which this unit is contained.
         */
        static createIn(container) {
            return internal.instancehelpers.createUnit(container, EntityChangeDataCaptureService);
        }
        /** @internal */
        _initializeDefaultProperties() {
            super._initializeDefaultProperties();
        }
    }
    EntityChangeDataCaptureService.structureTypeName = "ChangeDataCapture$EntityChangeDataCaptureService";
    EntityChangeDataCaptureService.versionInfo = new exports.StructureVersionInfo({
        introduced: "11.5.0",
        properties: {
            serviceName: {
                introduced: "11.8.0"
            }
        },
        experimental: {
            currentValue: true
        }
    }, internal.StructureType.ModelUnit);
    changedatacapture.EntityChangeDataCaptureService = EntityChangeDataCaptureService;
    /**
     * See: {@link https://docs.mendix.com/refguide/event-streams/#change-data-capture relevant section in reference guide}
     *
     * NOTE: This class is experimental and is subject to change in newer Model SDK versions.
     *
     * @ignore
     *
     * In version 11.5.0: introduced
     */
    class EntityChangeSource extends internal.Element {
        constructor(model, structureTypeName, id, isPartial, unit, container) {
            super(model, structureTypeName, id, isPartial, unit, container);
            /** @internal */
            this.__entity = new internal.ByNameReferenceProperty(EntityChangeSource, this, "entity", null, "DomainModels$Entity");
            /** @internal */
            this.__exposedName = new internal.PrimitiveProperty(EntityChangeSource, this, "exposedName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            this.__properties = new internal.PartListProperty(EntityChangeSource, this, "properties", []);
            if (arguments.length < 4) {
                throw new Error("new EntityChangeSource() cannot be invoked directly, please use 'model.changedatacapture.createEntityChangeSource()'");
            }
        }
        get containerAsEntityChangeDataCaptureService() {
            return super.getContainerAs(EntityChangeDataCaptureService);
        }
        get entity() {
            return this.__entity.get();
        }
        set entity(newValue) {
            this.__entity.set(newValue);
        }
        get entityQualifiedName() {
            return this.__entity.qualifiedName();
        }
        /**
         * In version 11.9.0: introduced
         */
        get exposedName() {
            return this.__exposedName.get();
        }
        set exposedName(newValue) {
            this.__exposedName.set(newValue);
        }
        /**
         * NOTE: This property is experimental and is subject to change in newer Model SDK versions.
         *
         * @ignore
         *
         * In version 11.9.0: introduced
         */
        get properties() {
            return this.__properties.get();
        }
        /**
         * Creates and returns a new EntityChangeSource instance in the SDK and on the server.
         * The new EntityChangeSource will be automatically stored in the 'sources' property
         * of the parent EntityChangeDataCaptureService element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  11.5.0 and higher
         */
        static createIn(container) {
            internal.createInVersionCheck(container.model, EntityChangeSource.structureTypeName, { start: "11.5.0" });
            return internal.instancehelpers.createElement(container, EntityChangeSource, "sources", true);
        }
        /**
         * Creates and returns a new EntityChangeSource instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        static create(model) {
            return internal.instancehelpers.createElement(model, EntityChangeSource);
        }
        /** @internal */
        _initializeDefaultProperties() {
            super._initializeDefaultProperties();
        }
    }
    EntityChangeSource.structureTypeName = "ChangeDataCapture$EntityChangeSource";
    EntityChangeSource.versionInfo = new exports.StructureVersionInfo({
        introduced: "11.5.0",
        properties: {
            entity: {
                required: {
                    currentValue: true
                }
            },
            exposedName: {
                introduced: "11.9.0"
            },
            properties: {
                introduced: "11.9.0"
            }
        },
        experimental: {
            currentValue: true
        }
    }, internal.StructureType.Element);
    changedatacapture.EntityChangeSource = EntityChangeSource;
})(changedatacapture = exports.changedatacapture || (exports.changedatacapture = {}));
